src/debug/profiler.rs

```diff
@@ -0,0 +1,183 @@
+    if auth.is_none() {
+        return (hyper::StatusCode::UNAUTHORIZED, "Unauthorized");
+    }
+
+    if let Ok(auth) = auth.unwrap().to_str() {
+        if auth != state.config.debug.secret {
+            return (hyper::StatusCode::UNAUTHORIZED, "Unauthorized");
+        }
+    }
+
+    tokio::spawn(run_profiler(payload));
+    (StatusCode::OK, "Profiler started.")
+}
+
+async fn run_profiler(payload: ProfilerPayload) {
+    let Ok(lock) = PROFILER_LOCK.try_lock() else {
+        tracing::warn!("profiler: failed to start because already running");
+        return;
+    };
+
+    if let Err(err) = profile(Duration::from_secs(payload.profile_duration)).await {
+        tracing::warn!(?err, "profiler: profiling failed");
+        return;
+    }
+
+    // Upload the profiles to S3.
+    upload_profiles().await;
+
+    drop(lock);
+}
+
+async fn profile(duration: Duration) -> anyhow::Result<()> {
+    let profiler = dhat::Profiler::new_heap();
+    tracing::warn!("profiler: begin profiling");
+
+    // Let the profiler run for specified duration.
+    tokio::time::sleep(duration).await;
+
+    drop(profiler);
+    tracing::warn!("profiler: end profiling");
+
+    Ok(())
+}
+
+async fn upload_profiles() {
+    tracing::warn!("profiler: uploading heap profiles in 30s...");
+
+    // Give some time to write profile data to disk if it's done on a different
+    // thread.
+    tokio::time::sleep(Duration::from_secs(30)).await;
+
+    let profiles_glob = "./dhat-*.json";
+
+    let Ok(file_list) = glob::glob(profiles_glob) else {
+        tracing::warn!("profiler: failed to read glob pattern");
+        return;
+    };
+
+    let ctx = {
+        let ctx = UPLOAD_CONTEXT.lock().await;
+
+        let Some(ctx) = ctx.as_ref() else {
+            tracing::warn!("profiler: upload context not initialized");
+            return;
+        };
+
+        ctx.clone()
+    };
+
+    tracing::warn!("profiler: uploading heap profiles");
+
+    let timestamp = chrono::Utc::now().timestamp_millis();
+    let key_prefix = format!("{timestamp}");
+
+    for path in file_list {
+        match path {
+            Ok(path) => {
+                if let Err(err) = upload_file(ctx.clone(), path.clone(), &key_prefix).await {
+                    tracing::warn!(?err, "profiler: failed to upload profile");
+                }
+
+                if let Err(err) = tokio::fs::remove_file(path).await {
+                    tracing::warn!(?err, "profiler: failed to remove profile after upload")
+                }
+            }
+
+            Err(err) => {
+                tracing::warn!(?err, "profiler: glob error");
+            }
+        }
+    }
+
+    tracing::warn!("profiler: finished uploading profiles");
+}
+
+async fn upload_file(ctx: UploadContext, path: PathBuf, key_prefix: &str) -> anyhow::Result<()> {
+    let Ok(metadata) = tokio::fs::metadata(&path).await else {
+        anyhow::bail!("failed to retrieve file metadata");
+    };
```

Comment:
Can't we use `?` here to preserve the original error?

```suggestion
    let metadata = tokio::fs::metadata(&path).await?;
```

Or maybe we need `.map_err()`:

```suggestion
    let metadata = tokio::fs::metadata(&path).await
        .map_err(|err| anyhow::anyhow!("failed to retrieve file metadata: {err:?}"))?;
```